### ISSUE TITLE
No Response body on error

### ISSUE THREAD

**Issue by roybrener** [square/retrofit#778]
I'm trying to understand what can be the cause for having no `Response` object on `failure

    @GET("/data")
    void getDatal(@Header(Config.AUTH_HEADER) String token, @Query("ts") long ts,  Callback<Data> callback);

    @Override
     public void failure (RetrofitError error) {
              LOGE(TAG, "@getData.failure > error: " + error);
              final Response response = error.getResponse();
                     if(response != null){
                               if(response.getStatus() == 401){
                                      doSomething();
                                }
                      } // response is null
      }

but Retrofit log is: 
    
    D/Retrofit﹕ WWW-Authenticate: Bearer realm="restservice", error="invalid_token: aaaaaa", error_description="Invalid access token:
    Content-Type: application/json;charset=UTF-8
     D/Retrofit﹕ Transfer-Encoding: chunked
     Date: Thu, 12 Mar 2015 09:31:58 GMT 03-12 11:31:56.397  30049-31480  D/Retrofit﹕ OkHttp-Selected-Protocol: http/1.1
     OkHttp-Sent-Millis: 1426152715315
     OkHttp-Received-Millis: 1426152716396
     D/Retrofit﹕ {"error":"invalid_token","error_description":"Invalid access token: aaaaaa}

**Comment by JakeWharton** [comment]
You didn't show the output of the log. Perhaps it's an unexpected error inside of Retrofit. Without more information there's not much to go on here...

**Comment by roybrener** [comment]
full log output:

    D/Retrofit﹕ <--- HTTP 401 http://10.0.0.2:8080/me/wall?ts=1426433499408 (22ms)
    D/Retrofit﹕ Server: Apache-Coyote/1.1
    D/Retrofit﹕ X-Content-Type-Options: nosniff
    D/Retrofit﹕ X-XSS-Protection: 1; mode=block
    D/Retrofit﹕ Cache-Control: no-cache, no-store, max-age=0, must-revalidate
    D/Retrofit﹕ Pragma: no-cache
    D/Retrofit﹕ Expires: 0
    D/Retrofit﹕ X-Frame-Options: DENY
    D/Retrofit﹕ Cache-Control: no-store
    D/Retrofit﹕ Pragma: no-cache
    D/Retrofit﹕ WWW-Authenticate: Bearer realm="restservice", error="invalid_token",     error_description="Invalid access token: aaaa"
    D/Retrofit﹕ Content-Type: application/json;charset=UTF-8
    D/Retrofit﹕ Transfer-Encoding: chunked
    D/Retrofit﹕ Date: Thu, 19 Mar 2015 12:58:22 GMT
    D/Retrofit﹕ OkHttp-Selected-Protocol: http/1.1
    D/Retrofit﹕ OkHttp-Sent-Millis: 1426769904583
    D/Retrofit﹕ OkHttp-Received-Millis: 1426769904590
    D/Retrofit﹕ {"error":"invalid_token","error_description":"Invalid access token: aaa"}
   D/Retrofit﹕ <--- END HTTP (106-byte body)

**Comment by roybrener** [comment]
any update on this issue?

**Comment by benzht** [comment]
Hi. I might have a similar issue. A slightly misbehaving server is sometimes replying with a String instead of the expected Integer. Subsequently, I get an

  retrofit.converter.ConversionException: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected an int but was STRING at line 1 column 1 path $

but the call to handleError(RetrofitError cause) has a null-cause. So far, I have not been able to discover a way to access the faulty string in the reply and work around the quirks of this server.

Any help would be very appreciated!

**Comment by JakeWharton** [comment]
@benzht You cannot. If you attempted to read the body as JSON first then the body has been exhausted and not buffered and therefore cannot be read.

**Comment by benzht** [comment]
Thanks for the fast reply, @JakeWharton 
Actually, it is not me reading the body of the response, but retrofit. This is my definition:
```
	@POST("/event")
	Integer sendEvent(@Body SensorReading reading) throws Exception;
```
Calling sendEvent sends the event to the server all right, during retrofit's processing of the response, the ConversionException appears and in my handleError I do not see what caused the problem or gives me the opportunity to do something reasonable.

**Comment by JakeWharton** [comment]
By using `Integer` as the response you are telling Gson that you want it to parse the response as an integer. However Gson's exception seems to indicate that it did not find an integer but a string in the HTTP response.

**Comment by benzht** [comment]
@JakeWharton, my issue/problem is not that the server's response provokes an error, but that I cannot seem to get hold of the (faulty) server's response in the handleError(RetrofitError cause). The variable _cause_ passed to handleError in the case outlined above does not seem to have any useful information.

**Comment by JakeWharton** [comment]
The request is not buffered so there will never be a way to access the actual data. Buffering the response would be extremely wasteful and slow for the success case (which is also the normal case).

**Comment by benzht** [comment]
OK, thank you for the explanation!

**Comment by roybrener** [comment]
@JakeWharton  can you please explain what you mean by 
> Buffering the response would be extremely wasteful and slow for the success case. 

even where there is an error on the server side, i'm always getting back from my server an `ErrorResponse` which i need to parse (for status code and message) . getBodyAs(ErrorResponse.class); does not produce anything. but i do get back `error.cause`

    com.example.DuplicateKeyException: retrofit.RetrofitError: 409 Conflict

So is there any way that i can get back the `ErrorResponse` object?

**Comment by JakeWharton** [comment]
@roybrener We have a test that verifies this works: https://github.com/square/retrofit/blob/2f4caa6f5df8e7f47a881266bbfc723f537b0b47/retrofit/src/test/java/retrofit/RestAdapterTest.java#L681-L699

Now I realize it's using `getBody()` not `getBodyAs()` but if you look at the implementation of `getBody()` you'll see that it just calls into `getBodyAs()` with the type from `getSuccessType()`: https://github.com/square/retrofit/blob/2f4caa6f5df8e7f47a881266bbfc723f537b0b47/retrofit/src/main/java/retrofit/RetrofitError.java#L111-L113

Without a failing test that demonstrates what you are seeing I cannot be of help. Please check out the `parent-1.9.0` branch for the v1.9.0 code and try to change the test to demonstrate what you are seeing. Please comment and I'll re-open, but closing for now.

**Comment by supunz** [comment]
@JakeWharton  Well I don't know whether this is the solution or not
i got same issue on when i used OkHttp client for the client of `RestAdapter` and it doesn't work correctly when I remove `restAdapter.setClient(new OkCliect(new OkHttpClient()));`
when i removed okhttp client library from my project it worked as it should.
I did it because "Retrofit will automatically use OkHttp (version 2.0 or newer)" when it is present.
[Proof](https://square.github.io/retrofit/)

**Comment by Krupen** [comment]
Solved it with following code in Retrofit 2.1.0. I also use okHttp3.

`InputStream i = response.errorBody().byteStream();
BufferedReader r = new BufferedReader(new InputStreamReader(i));
StringBuilder errorResult = new StringBuilder();
String line;
try {
while ((line = r.readLine()) != null) {
errorResult.append(line).append('\n');
}
} catch (IOException e) { 
e.printStackTrace(); 
}`
